BREAKING: Arrayable schemas, new Schema/* namespace, support schemas return raw arrays

edevsh opened the pull request:
## Summary

This pull request introduces full arrayable support across all schema-interface–based classes and ensures consistent test coverage for serialization behavior.

---

## Tasks

* [x] Implement `toArray()` method on `SchemaInteger`
* [x] Make all `SchemaInterface`-based classes implement arrayable behavior
* [x] (IN PROGRESS) Add and improve unit tests for all schema-interface–based classes to cover:
  * Field access and type validation
  * Array conversion (`toArray`) and consistency checks
  
---

## Breaking Changes

**1. `SchemaInterface` contract update**

* Added `SchemaInterface::toArray(): array`.
  → Any external implementations must add this method.

**2. Namespace & layout refactor**

* All schemas moved from `Blugen\Service\Lexicon\V1\TypeSpecificSchema\*` → `Blugen\Service\Lexicon\V1\Schema\*`.
* New structure:

  * **Concrete** → primitive field schemas (e.g., `StringSchema`, `BooleanSchema`)
  * **Container** → structural types (`ArraySchema`, `ObjectSchema`)
  * **Meta** → compositional/indirect types (`RefSchema`, `UnionSchema`, `UnknownSchema`, `TokenSchema`)
  * **Support** → auxiliary schemas (`InputSchema`, `OutputSchema`, `MessageSchema`, `ErrorsSchema`)
  * **Primary** → root schemas (`QuerySchema`, `ProcedureSchema`, `RecordSchema`, `SubscriptionSchema`, `HTTPAbstract`)
    → Any code importing old `TypeSpecificSchema\...` namespaces will break.

**3. Support schemas – `type()` and `description()` behavior**

* Now use `SupportSchemaTrait`, which throws `BadMethodCallException` in `type()` (and `description()` unless overridden).
* Previously, these returned a concrete type string.
  → Code calling `type()` on support schemas will fail.

**4. InputSchema / OutputSchema return type changes**

* `schema()` now returns raw `?array` from `IOAbstract` (was schema object).
* Default empty schema behavior removed.
  → Callers expecting schema objects (e.g., `ObjectSchema`, `UnionSchema`, `RefSchema`) will break.

**5. `ErrorsSchema` collection semantics**

* Now yields `ErrorSchema` objects instead of arrays.
* Implements read-only `ArrayAccess`:

  * `offsetSet()` / `offsetUnset()` throw `BadMethodCallException`.
  * `offsetGet()` throws `OutOfBoundsException` for invalid indexes.
    → Previously it accepted arbitrary array access.

**6. `ObjectSchema::properties()` contract update**

* Return type changed: now returns an array of schema instances (via `SchemaFactory`) instead of `Property` objects.
* Throws `MissingRequiredFieldException` if `properties` is invalid or missing.
* Property metadata (e.g. `name`, `nullable`, `required`) must now be derived via schema arrays.
  → Code depending on `Property` objects will break.

**7. New core exception introduced**

* `Blugen\Service\Lexicon\V1\Exceptions\MissingRequiredFieldException`
  → Thrown in `ObjectSchema::properties()` and potentially other locations.

**8. Generator & factory updates**

* `SchemaFactory` refactored to point to new `Schema\{Concrete,Container,Meta}` classes.
* Generators now build schemas via `SchemaFactory`.
  → Any custom generator or class references to old `TypeSpecificSchema` paths must be updated.

**9. Legacy class removals**

* Deleted:

  * `V1/TypeSpecificSchema/Support/{ErrorsSchema,InputSchema,MessageSchema,OutputSchema}`
  * `V1/TypeSpecificSchema/Field/ObjectSchema`
* Replacements exist under new `V1/Schema/{Support,Container,...}` but with changed behaviors above.

**10. Explicit breaking commit reference**

* `16a45c4` — **BREAKING CHANGE:** Refactor(TypeSpecificSchema/Support) to use traits and return raw schema arrays.

shahmal1yev commented:
### Example

#### From
<details>

<summary>Click Here</summary>

```json
{
  "lexicon": 1,
  "id": "app.test.profile",
  "defs": {
    "main": {
      "type": "object",
      "description": "Main profile object",
      "properties": {
        "displayName": {
          "type": "string",
          "description": "User's display name",
          "minLength": 3,
          "maxLength": 50,
          "default": "Anonymous"
        },
        "bio": {
          "type": "string",
          "description": "Short biography of the user",
          "maxGraphemes": 200,
          "knownValues": ["developer", "designer", "manager"]
        },
        "email": {
          "type": "string",
          "description": "User's email address",
          "format": "at-identifier",
          "const": "user@example.com"
        },
        "age": {
          "type": "integer",
          "description": "User's age",
          "minimum": 18,
          "maximum": 120
        },
        "score": {
          "type": "integer",
          "description": "Reputation score",
          "enum": [0, 10, 20, 30],
          "default": 10
        },
        "isActive": {
          "type": "boolean",
          "description": "Is the profile currently active?",
          "default": true
        },
        "isBanned": {
          "type": "boolean",
          "description": "Has the user been banned?",
          "const": false
        },
        "settings": {
          "type": "object",
          "description": "User settings",
          "properties": {
            "newsletterSubscribed": {
              "type": "boolean",
              "description": "Subscribed to newsletter",
              "default": false
            },
            "timezoneOffset": {
              "type": "integer",
              "description": "Timezone offset in minutes",
              "minimum": -720,
              "maximum": 840
            },
            "signature": {
              "type": "string",
              "description": "Optional signature text",
              "minLength": 0,
              "maxLength": 140,
              "default": ""
            }
          },
          "required": ["timezoneOffset"],
          "nullable": ["signature"]
        }
      },
      "required": ["displayName", "age", "isActive"]
    }
  }
}

```

</details>

#### To

<details>

<summary>Click Here</summary>

```php
<?php

declare(strict_types=1);

namespace App\Test;

class Profile
{
	/**
	 * @var string
	 * @length 3-50
	 * @default Anonymous
	 *
	 * User's display name
	 */
	private string $displayName;

	/**
	 * @var string
	 * @knownValues {developer, designer, manager}
	 * @graphemes 0-200
	 *
	 * Short biography of the user
	 */
	private string $bio;

	/**
	 * @var string
	 * @format at-identifier
	 * @const user@example.com
	 *
	 * User's email address
	 */
	private string $email;

	/**
	 * @var int
	 * @range 18 to 120
	 *
	 * User's age
	 */
	private int $age;

	/**
	 * @var int
	 * @enum {0, 10, 20, 30}
	 * @default 10
	 *
	 * Reputation score
	 */
	private int $score;

	/**
	 * @var bool
	 *
	 * Is the profile currently active?
	 * @default true
	 */
	private bool $isActive;

	/**
	 * @var bool
	 *
	 * Has the user been banned?
	 * @const false
	 */
	private bool $isBanned;
	private object $settings;


	/**
	 * Get the value of $displayName.
	 *
	 * @return string
	 */
	public function getDisplayName(): string
	{
		return $this->displayName;
	}


	/**
	 * Set the value of $displayName
	 *
	 * @param string $displayName
	 * @return self
	 */
	public function setDisplayName(string $displayName): self
	{
		$this->displayName = $displayName ?? 'Anonymous';
		return $this;
	}


	/**
	 * Get the value of $bio.
	 *
	 * @return string
	 */
	public function getBio(): string
	{
		return $this->bio;
	}


	/**
	 * Set the value of $bio
	 *
	 * @param string $bio
	 * @return self
	 */
	public function setBio(string $bio): self
	{
		$this->bio = $bio;
		return $this;
	}


	/**
	 * Get the value of $email.
	 *
	 * @return string
	 */
	public function getEmail(): string
	{
		return $this->email;
	}


	/**
	 * Set the value of $email
	 *
	 * @param string $email
	 * @note This value is constant and cannot be changed.
	 * @return self
	 */
	public function setEmail(string $email): self
	{
		// Const field. This value is fixed.
		return $this;
	}


	/**
	 * Get the value of $age.
	 *
	 * @return int
	 */
	public function getAge(): int
	{
		return $this->age;
	}


	/**
	 * Set the value of $age
	 *
	 * @param int $age
	 * @return self
	 */
	public function setAge(int $age): self
	{
		$this->age = $age;
		return $this;
	}


	/**
	 * Get the value of $score.
	 *
	 * @return int
	 */
	public function getScore(): int
	{
		return $this->score;
	}


	/**
	 * Set the value of $score
	 *
	 * @param int $score
	 * @return self
	 */
	public function setScore(int $score): self
	{
		if (!in_array($score, [0, 10, 20, 30], true)) {
		    throw new \InvalidArgumentException('Invalid value for score');
		}
		$this->score = $score ?? 10;
		return $this;
	}


	/**
	 * Get the value of $isActive.
	 *
	 * @return bool
	 */
	public function getIsActive(): bool
	{
		return $this->isActive;
	}


	/**
	 * Set the value of $isActive
	 *
	 * @param bool $isActive
	 * @return self
	 */
	public function setIsActive(bool $isActive): self
	{
		$this->isActive = $isActive ?? true;
		return $this;
	}


	/**
	 * Get the value of $isBanned.
	 *
	 * @return bool
	 */
	public function getIsBanned(): bool
	{
		return $this->isBanned;
	}


	/**
	 * Set the value of $isBanned
	 *
	 * @param bool $isBanned
	 * @note This value is constant and cannot be changed.
	 * @return self
	 */
	public function setIsBanned(bool $isBanned): self
	{
		// Const field. This value is fixed.
		return $this;
	}


	public function __construct()
	{
		$this->settings = new class {
			/**
			 * @var bool
			 *
			 * Subscribed to newsletter
			 * @default false
			 */
			private bool $newsletterSubscribed;

			/**
			 * @var int
			 * @range -720 to 840
			 *
			 * Timezone offset in minutes
			 */
			private int $timezoneOffset;

			/**
			 * @var string|null
			 * @length 0-140
			 *
			 * Optional signature text
			 */
			private ?string $signature;


			/**
			 * Get the value of $newsletterSubscribed.
			 *
			 * @return bool
			 */
			public function getNewsletterSubscribed(): bool
			{
				return $this->newsletterSubscribed;
			}


			/**
			 * Set the value of $newsletterSubscribed
			 *
			 * @param bool $newsletterSubscribed
			 * @return self
			 */
			public function setNewsletterSubscribed(bool $newsletterSubscribed): self
			{
				$this->newsletterSubscribed = $newsletterSubscribed ?? false;
				return $this;
			}


			/**
			 * Get the value of $timezoneOffset.
			 *
			 * @return int
			 */
			public function getTimezoneOffset(): int
			{
				return $this->timezoneOffset;
			}


			/**
			 * Set the value of $timezoneOffset
			 *
			 * @param int $timezoneOffset
			 * @return self
			 */
			public function setTimezoneOffset(int $timezoneOffset): self
			{
				$this->timezoneOffset = $timezoneOffset;
				return $this;
			}


			/**
			 * Get the value of $signature.
			 *
			 * @return string|null
			 */
			public function getSignature(): ?string
			{
				return $this->signature;
			}


			/**
			 * Set the value of $signature
			 *
			 * @param string|null $signature
			 * @return self
			 */
			public function setSignature(?string $signature): self
			{
				$this->signature = $signature;
				return $this;
			}
		};
	}


	public function getSettings(): object
	{
		return $this->settings;
	}
}
```

</details>

edevsh committed:
Refactor(SchemaInterface): Add `toArray` method to SchemaInterface

edevsh committed:
Add arrayable schema/definition traits

edevsh committed:
Refactor(Lexicon/V1): integrate traits and centralize schema access

- Schema: use ArrayableTrait and SchemaTrait
- Definition: add schema() and toArray() (filters nulls)

edevsh committed:
Refactor(Lexicon/V1/TypeSpecificSchema): apply ArrayableTrait and SchemaTrait

- Add trait imports and use to all field schema classes
- Standardize array serialization and centralize schema access across fields

edevsh committed:
Refactor(TypeSpecificDefinition/Primary): apply ArrayableTrait and DefinitionTrait

- Add trait imports and use to Procedure, Query, Record, Subscription definitions
- Standardize array serialization and shared definition helpers (no behavior change)

edevsh committed:
Refactor(TypeSpecificDefinition/Field): apply ArrayableTrait and DefinitionTrait

- Import traits and add use in Array, Object, String field definitions
- Standardize array serialization and shared definition helpers (no behavior change)

edevsh committed:
BREAKING CHANGE: Refactor(TypeSpecificSchema/Support): use traits; return raw schema arrays

- Apply ArrayableTrait and SchemaTrait to Errors, Input, Message, Output
- Change schema() to return underlying arrays instead of wrapper instances
- Add ErrorsSchema::schema(); simplify defaults and defer wrapping to callers

edevsh committed:
Refactor(DefGenerator/Procedure): build schema via SchemaFactory

- Import SchemaFactory and construct schema from input array
- Replace direct wrapper access with factory-based instantiation
- Prepares generator for new array-returning schema() behavior

edevsh committed:
Add unit tests for ArrayableTrait

edevsh committed:
Add unit tests for SchemaTrait

edevsh committed:
Add unit tests for DefinitionTrait

edevsh committed:
Refactor(Support/OutputSchema): Handle optional `schema` field in query/primary schema definitions

edevsh committed:
Test(TypeSpecificSchema/Field): Add unit tests

edevsh committed:
Test(Input/Output): Add unit tests for Input and Output schemas

edevsh committed:
Chore(IOAbstract): Add an abstract class to apply dry on Input and Output schemas

edevsh committed:
Refactor(Input/Output): Update Input and Output support schemas based on requirements

edevsh committed:
Test(Support/MessageSchema): Add unit tests for `MessageSchema`

edevsh committed:
Fix(Support/MessageSchema): Update `MessageSchema` based on failed tests and optimize imports

edevsh committed:
Chore(Trait/SupportSchema): Introduce \`SupportSchemaTrait\` with explicit BadMethodCallException for unsupported methods

edevsh committed:
Test(Traits): Add unit test traits for schema behavior and support schema validation

edevsh committed:
Test(Support/ErrorSchema): Add dedicated ErrorSchemaTest overriding trait behavior for description handling

edevsh committed:
Test(Support/ErrorsSchema): Add comprehensive unit tests for ErrorsSchema iterable and read-only behaviors

edevsh committed:
Refactor(Support/ErrorsSchema): Replace manual error mapping with ErrorSchema collection and add read-only ArrayAccess support

edevsh committed:
Feat(Support/ErrorSchema): Introduce ErrorSchema wrapper for individual error definitions with schema delegation

edevsh committed:
Refactor(Support/InputSchema): Simplify IOAbstract delegation and update InputSchema tests for description and schema accessors

edevsh committed:
Refactor(Lexicon/OutputSchema): Simplify OutputSchema by delegating to IOAbstract and align tests with unified schema behavior

edevsh committed:
Refactor(Lexicon/MessageSchema): Adopt SupportSchemaTrait and align tests with shared trait-based schema validation

edevsh committed:
Fix(Tests/Traits & ErrorsSchema): Resolve missing trait class references and improve offset exception handling

- Rename and update all reusable test traits to use `*Trait` suffix:
  • WithArrayableTest → WithArrayableTestTrait
  • WithGetTest → WithGetTestTrait
  • WithSchemaTest → WithSchemaTestTrait
  • WithSupportSchemaTest → WithSupportSchemaTestTrait
- Update all schema test classes (Input, Output, Error, Errors, Message) to reference renamed traits
- Fix PHPUnit warnings caused by outdated trait names
- Update ErrorsSchema::offsetGet() to throw OutOfBoundsException with clear message instead of implicit TypeError
- Adjust related test to assert explicit exception type and message

edevsh committed:
Test(Schema/Primary): Add unit tests for primary type schemas

edevsh committed:
Feat(Schema/Primary): Implement primary type schemas based on tests

edevsh committed:
Chore: Introduce `DefaultSchemaTrait` with default accessors for all schemas to improve DRY

edevsh committed:
Refactor(SchemaTrait): Rename `SchemaTrait` as `RawSchemaAccessorTrait`

edevsh committed:
Refactor(DefaultSchemaTrait): Rename `DefaultSchemaTrait` as `SchemaTrait`

edevsh committed:
Chore: Fix imports

edevsh committed:
Rename TypeSpecificSchema as Schema

edevsh committed:
Move Concrete types to Concrete namespace

edevsh committed:
Move container types to Container namespace

edevsh committed:
Move meta types to Meta namespace

edevsh committed:
Rewrite ObjectSchema unit tests

edevsh committed:
Fix issues on ObjectSchema based on refactored test classs

edevsh committed:
Introduce MissingRequiredFieldException

edevsh committed:
Fix Input/Output schema tests for `schema` field: it is not required

edevsh committed:
Backward compatibility

edevsh committed:
Fix MessageSchema::schame() return type

edevsh committed:
Introduce CHANGELOG.md
